src/verifast1.ml

```diff
@@ -2520,13 +2522,43 @@ module VerifyProgram1(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
     | ContainsAnyConstraint allowContainsAnyPositive -> type_satisfies_contains_any_constraint allowContainsAnyPositive tp
 
   (* Region: type compatibility checker *)
+
+  let direct_base_addr (derived_name, derived_addr) base_name =
+    let _, Some (bases, _), _, _ = List.assoc derived_name structmap in
+    let _, _, base_offset = List.assoc base_name bases in
+    ctxt#mk_add derived_addr base_offset
+
+  let base_addr l (derived_name, derived_addr) base_name =
+    let rec iter derived_name offsets =
+      let _, Some (bases, _), _, _ = List.assoc derived_name structmap in 
+      let other_paths = bases |> List.fold_left begin fun acc (name, (_, _, offset)) -> 
+        match iter name (offset :: offsets) with
+        | Some p -> p :: acc
+        | None -> acc
+      end [] in
+      match try_assoc base_name bases, other_paths with 
+      | Some _,  _ :: _
+      | None, _ :: (_ :: _) -> static_error l (Printf.sprintf "Derived '%s' to base '%s' is ambiguous: multiple '%s' base candidates exist." derived_name base_name base_name) None
+      | Some (_, _, base_offset), [] -> Some (base_offset :: offsets)
+      | None, [p] -> Some p
+      | _ -> None
+    in 
+    let Some offsets = iter derived_name [] in 
+    List.rev offsets |> List.fold_left ctxt#mk_add derived_addr
+
+  let rec is_derived_of_base derived_name base_name =
+    let check_bases bases = bases |> List.exists @@ fun (name, _) -> is_derived_of_base name base_name in
+    derived_name = base_name ||
+    match try_assoc derived_name structmap with 
+    | Some (_, Some (bases, _), _, _) -> check_bases bases 
+    | None -> false
   
   let rec compatible_pointees t t0 =
     match (t, t0) with
       (_, Void) -> true
     | (Void, _) -> true
     | (PtrType t, PtrType t0) -> compatible_pointees t t0
-    | StructType x, StructType y when dialect = Some Cxx && is_derived_struct_of x y -> true
+    | StructType x, StructType y when dialect = Some Cxx && is_derived_of_base x y -> true

```

Comment:
This seems wrong. Actually, the old version seems wrong as well, but this seems even more wrong. The old version was already wrong because `compatible_pointees` and `unify` are supposed to be symmetric: `unify t1 t2` should hold if and only if `unify t2 t1` holds. Another reason why the old version was wrong is that it is dangerous to allow an implicit conversion from `Derived**` to `Base**`, because this suggests incorrectly that the pointer points to a variable into which any value of type `Base*` can be stored. Technically speaking: pointers are invariant, no covariant (and not contravariant).

The new version is even more wrong because implicitly converting a `Derived**` to a `Base**` where the offset of `Base` in `Derived` is nonzero suggests incorrectly that dereferencing that pointer will yield a valid pointer to an object of type `Base`.

Implicit widening conversions, like from `Derived` to `Base`, should be enabled in `expect_type_core`, not in `unify`.

(Not that any of this matters for soundness; in VeriFast, all pointer types really mean the same thing. But if we allow this, we might as well allow implicit conversion from any `Foo*` to any `Bar*`.)